incorrect diagosis of default template argument for function declaration inside class template member function

 http://gcc.gnu.org/bugzilla/show_bug.cgi?id=60605

--- gcc-4_8-branch/gcc/testsuite/g++.dg/template/local-fn1.C	(revision 0)
+++ gcc-4_8-branch/gcc/testsuite/g++.dg/template/local-fn1.C	(revision 211753)
@@ -0,0 +1,8 @@
+// PR c++/60605
+
+template <typename T = int>
+struct Foo {
+    void bar() {
+        void bug();
+    }
+};
--- gcc-4_8-branch/gcc/cp/ChangeLog	(revision 211752)
+++ gcc-4_8-branch/gcc/cp/ChangeLog	(revision 211753)
@@ -1,3 +1,8 @@
+2014-06-17  Jason Merrill  <jason@redhat.com>
+
+	PR c++/60605
+	* pt.c (check_default_tmpl_args): Check DECL_LOCAL_FUNCTION_P.
+
 2014-06-02  Jason Merrill  <jason@redhat.com>
 
 	PR c++/61134
--- gcc-4_8-branch/gcc/cp/pt.c	(revision 211752)
+++ gcc-4_8-branch/gcc/cp/pt.c	(revision 211753)
@@ -4308,7 +4308,8 @@ check_default_tmpl_args (tree decl, tree parms, bo
      in the template-parameter-list of the definition of a member of a
      class template.  */
 
-  if (TREE_CODE (CP_DECL_CONTEXT (decl)) == FUNCTION_DECL)
+  if (TREE_CODE (CP_DECL_CONTEXT (decl)) == FUNCTION_DECL
+      || (TREE_CODE (decl) == FUNCTION_DECL && DECL_LOCAL_FUNCTION_P (decl)))
     /* You can't have a function template declaration in a local
        scope, nor you can you define a member of a class template in a
        local scope.  */
